File: _randomqueen/sann/vezer.jl
using StatsBase
using Printf

  mutable struct state
    N::Int
    x::Vector{Int}
    poz::Vector{Int}
    neg::Vector{Int}
    E::Int
    function state(x::Vector{Int})
      N=length(x)
      poz=fill(0,2N)
      neg=fill(0,2N)
      for i in 1:N
        poz[i+x[i]]+=1
        neg[i-x[i]+N]+=1
      end
      E=0
      for t in 1:2N
        (poz[t]>1) && (E+=poz[t])
        (neg[t]>1) && (E+=neg[t])
      end
      new(N,copy(x),poz,neg,E)
    end
    state(s::state)=new(s.N,copy(s.x),copy(s.poz),copy(s.neg),s.E)

  end

function nqueen_sann(N::Int)

  function deltaE(s::state,I)
    i,j=I
    xi,xj=s.x[i],s.x[j]

    s.poz[i+xi]-=1
    s.poz[j+xj]-=1
    s.neg[i+N-xi]-=1
    s.neg[j+N-xj]-=1
    s.poz[i+xj]+=1
    s.poz[j+xi]+=1
    s.neg[i+N-xj]+=1
    s.neg[j+N-xi]+=1

    E=sum(s.poz.*(s.poz.>1))+sum(s.neg.*(s.neg.>1))

    s.poz[i+xi]+=1
    s.poz[j+xj]+=1
    s.neg[i+N-xi]+=1
    s.neg[j+N-xj]+=1
    s.poz[i+xj]-=1
    s.poz[j+xi]-=1
    s.neg[i+N-xj]-=1
    s.neg[j+N-xi]-=1

    E-s.E
    
  end

  function move(s::state,I,dE)
    s.E+=dE
    i,j=I
    xi,xj=s.x[i],s.x[j]

    s.poz[i+xi]-=1
    s.poz[j+xj]-=1
    s.neg[i+N-xi]-=1
    s.neg[j+N-xj]-=1
    s.poz[i+xj]+=1
    s.poz[j+xi]+=1
    s.neg[i+N-xj]+=1
    s.neg[j+N-xi]+=1

    s.x[i],s.x[j]=s.x[j],s.x[i]
  end



  mE=0
  i=0
  while i<N
    s=state(sample(1:N,N; replace=false))
    I=sample(1:N,2;replace=false)
    dE=deltaE(s,I)
    if dE>=0
      mE+=dE
      i+=1
    end    
  end
  mE=mE/N
  T0=-mE/log(0.5)

  s=state(sample(1:N,N; replace=false))

  xbest=copy(s.x)
  Ebest=s.E


println(T0)  
  T1=1e-50
  alfa0=0.8
  alfa1=0.9999
  #alfa0=alfa1

  T=T0
  alfa=alfa0

  maxstep=10*N^2
  step=0
  while T>T1 && Ebest>0 && step<maxstep
    step=step+1
    I=sample(1:N,2;replace=false)
    dE=deltaE(s,I)
    if dE<0
      move(s,I,dE)
      if s.E<Ebest
        xbest.=s.x
        Ebest=s.E
      end
    else
      if rand()<exp(-dE/T)
        move(s,I,dE)
      end
    end

    T=alfa*T
    alfa=alfa0+(T-T0)*(alfa1-alfa0)/(T1-T0)
  end




  @printf("Ebest=%d step=%d T=%.2e\n",Ebest,step,T)

  xbest

end

File: _randomqueen/sann/vezer.m
clc;
N=100;
[x,hx]=szh(N);
hx
% xx=zeros(N,N);
% for i=1:N
%     xx(i,x(i))=1;
% end
% imagesc(xx)
% axis square


function [x,hx]=szh(N)
    poz=zeros(2*N,1);
    neg=zeros(2*N,1);

    x=randsample(N,N);
    hx=h(x);

    T0=1;
    T1=1e-22;
    alfa0=0.8;
    alfa1=0.9999;

    T=T0;
    alfa=alfa0;
    %update=@(T,alfa,i) 0.9999*T;
    %update=@(T,i) T0/(1+i)^4;

    

    step=0;
    while T>T1 && hx>0
        step=step+1;
        y=xnew(x);
        hy=h(y);
        dh=hy-hx;
        if dh<0
            x=y;
            hx=hy;
        else
            if rand()<exp(-dh/T)
                x=y;
                hx=hy;
            end
        end
        
        T=alfa*T;
        alfa=alfa0+(T-T0)*(alfa1-alfa0)/(T1-T0);
    end

    function x=xnew(x)
        I=randsample(N,2);
        x(I)=x([I(2);I(1)]);
        % if rand()<0.5
        %     I=randsample(N,2);
        %     x(I)=x([I(2);I(1)]);
        % end
        
    end

    function hx=h(x)
        % poz=zeros(2*N,1);
        % neg=zeros(2*N,1);
        poz(:)=0;
        neg(:)=0;
        for i=1:N
            t=i+x(i);
            poz(t)=poz(t)+1;
            t=i-x(i)+N;
            neg(t)=neg(t)+1;
        end
        %hx=sum(poz>1)+sum(neg>1);
        hx=sum(poz.*(poz>1))+sum(neg.*(neg>1));
        % hx=0;
        % for i=1:2*N
        %   hx=hx+(poz(i)>1).*poz(i);
        %   poz(i)=0;
        %   hx=hx+(neg(i)>1).*neg(i);
        %   neg(i)=0;
        % end
        
    end

    fprintf("step=%d T=%.2e\n",step,T);

end